src/ucs/sys/init.c

```diff
@@ -13,8 +14,67 @@
 #include <ucs/async/async.h>
 
 
+/* run-time CPU detection */
+static void UCS_F_NOOPTIMIZE ucs_cpu_detection(void)
+{
+    char str[256];
+    char *p_str;
+    int cpu_flags;
+    struct {
+        const char* flag;
+        ucs_cpu_flag_t value;
+    } *p_flags,
+    cpu_flags_array[] = {
+            { "cmov", UCS_CPU_FLAG_CMOV },
+            { "mmx", UCS_CPU_FLAG_MMX },
+            { "mmx2", UCS_CPU_FLAG_MMX2 },
+            { "sse", UCS_CPU_FLAG_SSE },
+            { "sse2", UCS_CPU_FLAG_SSE2 },
+            { "sse3", UCS_CPU_FLAG_SSE3 },
+            { "ssse3", UCS_CPU_FLAG_SSSE3 },
+            { "sse41", UCS_CPU_FLAG_SSE41 },
+            { "sse42", UCS_CPU_FLAG_SSE42 },
+            { "avx", UCS_CPU_FLAG_AVX },
+            { "avx2", UCS_CPU_FLAG_AVX2 },
+            { NULL, UCS_CPU_FLAG_UNKNOWN },
+    };
+
+    cpu_flags = ucs_arch_get_cpu_flag();
+    if (UCS_CPU_FLAG_UNKNOWN == cpu_flags) {
+        return ;
+    }
+#ifdef CPU_FLAGS
+    strncpy(str, UCS_PP_STR(CPU_FLAGS), sizeof(str) - 1);
+#else
+    strncpy(str, "", sizeof(str) - 1);
+#endif
+
+    p_str = strtok(str, " |\t\n\r");
+    while (p_str) {
+        p_flags = cpu_flags_array;
+        while (p_flags && p_flags->flag) {
```

Comment:
instead of doing it in O(n^2), go over the array of existing cpu flags and print error if a flag is required but not supported by the cpu. this will be simpler. in addition, you may go over CPU_FLAGS and make sure it is contained in cpu_flags_array, e.g it has a valid value.